cli/command/trust/sign.go

```
@@ -52,10 +61,45 @@ func signImage(cli command.Cli, image, tag string) error {
 			return trust.NotaryError(repoInfo.Name.Name(), err)
 		}
 	}
-
-	// TODO: craft and sign the target using image.addTargetToAllSignableRoles (needs to be exported)
+	requestPrivilege := command.RegistryAuthenticationPrivilegedFunc(cli, repoInfo.Index, "push")
+	target, err := createTarget(notaryRepo, tag)
+	if err != nil && strings.Contains(err.Error(), "No valid trust data for") {
+		return image.TrustedPush(ctx, cli, repoInfo, ref, *authConfig, requestPrivilege)
+	} else if err != nil {
+		return err
+	}
+	err = image.AddTargetToAllSignableRoles(notaryRepo, &target)
+	if err == nil {
+		err = notaryRepo.Publish()
+	}
+	if err != nil {
+		fmt.Fprintf(cli.Out(), "Failed to sign %q:%s - %s\n", repoInfo.Name.Name(), tag, err.Error())
+		return trust.NotaryError(repoInfo.Name.Name(), err)
+	}
+	fmt.Fprintf(cli.Out(), "Successfully signed %q:%s\n", repoInfo.Name.Name(), tag)
 	return nil
 }
+func createTarget(notaryRepo *client.NotaryRepository, tag string) (client.Target, error) {
+	target := &client.Target{}
+	var err error
+	if tag == "" {
+		return *target, fmt.Errorf("No tag specified")
+	}
+	target.Name = tag
+	target.Hashes, target.Length, err = getSignedManifestHashAndSize(notaryRepo, tag)
+	return *target, err
+}
+func getSignedManifestHashAndSize(notaryRepo *client.NotaryRepository, tag string) (data.Hashes, int64, error) {
+	targets, err := notaryRepo.GetAllTargetMetadataByName(tag)
+	if err == nil {
+		for _, tgt := range targets {
+			if isReleasedTarget(tgt.Role.Name) {
+				return tgt.Target.Hashes, tgt.Target.Length, nil
+			}
+		}
+	}
+	return nil, 0, err
```

we should return an explicit error here, `err` could be `nil`